model/request/ProctorioRequestBuilder.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA
+ */
+
+namespace oat\remoteProctoring\model\request;
+
+use common_exception_Error;
+use common_exception_NotFound;
+use core_kernel_classes_Resource;
+use oat\generis\Helper\UuidPrimaryKeyTrait;
+use oat\oatbox\log\LoggerAwareTrait;
+use oat\oatbox\user\User;
+use oat\Proctorio\ProctorioConfig;
+use oat\remoteProctoring\model\ProctorioApiService;
+use oat\tao\helpers\UserHelper;
+use oat\taoDelivery\model\execution\DeliveryExecutionInterface;
+use tao_helpers_Uri;
+use Throwable;
+
+class ProctorioRequestBuilder
+{
+    use UuidPrimaryKeyTrait;
+    use LoggerAwareTrait;
+
+    /** @var array $options */
+    private $options;
+
+    /**
+     * @param DeliveryExecutionInterface $deliveryExecution
+     * @param string $launchUrl
+     * @param array $options
+     * @return array
+     * @throws common_exception_Error
+     * @throws common_exception_NotFound
+     */
+    public function build(DeliveryExecutionInterface $deliveryExecution, string $launchUrl, array $options): array
+    {
+        $this->options = $options;
+
+        return
+            [
+                //delivery execution level
+                ProctorioConfig::LAUNCH_URL => $launchUrl,
+                ProctorioConfig::USER_ID => $deliveryExecution->getUserIdentifier(),
+
+                //platform level
+                ProctorioConfig::OAUTH_CONSUMER_KEY => $this->getOauthCredentials(),
+
+                ProctorioConfig::EXAM_START => $launchUrl,
+                ProctorioConfig::EXAM_TAKE => $this->getExamUrl(),
+                ProctorioConfig::EXAM_END => $this->getExamUrl(),
+                ProctorioConfig::EXAM_SETTINGS => $this->getExamSettings(),
+
+                //delivery execution level
+                ProctorioConfig::FULL_NAME => $this->getUserFullName($deliveryExecution),
+                //Delivery level
+                ProctorioConfig::EXAM_TAG => $deliveryExecution->getDelivery()->getLabel(),
+
+                ProctorioConfig::OAUTH_TIMESTAMP => $this->getTime(),
+                ProctorioConfig::OAUTH_NONCE => $this->getNonce(),
+            ];
+    }
+
+
+    /**
+     * @param $name
+     * @return mixed|null
+     */
+    private function getOption($name)
```

Comment:
Might be something like : 
```suggestion
    private function getOption(string $name): ?string
```

Reply:
the original one has a mixed possibility, and because we might use something like an array inside the config , i am reluctant to use string as a type hint or as response type hint

Reply:
What is:
> original one
Can we ensure that we get string param and string return?
Is it possible? We should have here type hinting everywhere, at all cost.

Reply:
i don't think that's the case